default/python/AI/colonization/calculate_planet_colonization_rating.py

```diff
@@ -1402,27 +1402,36 @@ def _adapt_supply_factor(supply_factor: float, species: fo.species) -> float:
         return supply_factor
 
 
-def _colony_upkeep(num_exobots: int, num_normal: int, num_outpost: int) -> float:
+def _colony_upkeep(num_exobots: int, num_normal: int, num_outpost: int, num_planets_in_galaxy: int) -> float:
     colonies = num_exobots + num_normal - 1  # assume we have a capital, calculation won't be too wrong otherwise
-    per_colony = get_named_real("COLONY_ADMIN_COSTS_PER_PLANET")
+    influence_focused_ratio = get_named_real("END_GAME_PROPORTION_INFLUENCE_FOCUSED")
+    IP_per_colony = get_named_real("END_GAME_COLONY_IP_OUTPUT")
     outpost_factor = get_named_real("OUTPOST_RELATIVE_ADMIN_COUNT")
+    ratio_owned_planets = (num_normal + outpost_factor * (num_exobots + num_outpost)) / num_planets_in_galaxy
     # see influence.macros, exobot colonies pay like a colony, but count as outposts for the cost increase
-    return colonies * per_colony * (num_normal + outpost_factor * (num_exobots + num_outpost)) ** 0.5
+    return (
+        colonies
+        * influence_focused_ratio
+        * IP_per_colony
+        / (1 + 2.71828182846 ** (-10 * (ratio_owned_planets - 2 / 3)))
+    )
 
 
 def _rate_upkeep(planet: fo.planet, species_name: SpeciesName, details: list) -> float:
     exobot = SpeciesName("SP_EXOBOT")
     num_exobots = len(get_empire_planets_by_species().get(exobot, []))
     num_normal = len(get_empire_populated_planets()) - num_exobots
     num_outpost = len(get_empire_outposts())
-    current_upkeep = _colony_upkeep(num_exobots, num_normal, num_outpost)
+    universe = fo.getUniverse()
+    num_planets_in_galaxy = len(universe.planetIDs)
```

Comment:
I have no idea how to tackle this.

Reply:
Is current number of system known to all clients? Value shown in galaxy settings could be different from it.

Reply:
> Is current number of system known to all clients?

not trivially / explicitly. possibly it's determinable by some unintended roundabout calcuation based on something observable.

Reply:
Ok, tested for 50, 150 and 450 planets, three times each. I got
Low density: systems * [0.88-0.93].
Medium density: systems * [1.17-1.24].
High density: systems * [1.75-1.9]. (I got three times exactly 385 planets for the 150 systems galaxy).

So I could use galaxy_size * [0.9 if low, 1.2 if medium, 1.8 if high]

Reply:
> So I could use galaxy_size * [0.9 if low, 1.2 if medium, 1.8 if high]

3 samples is a bit low, I did a few more with 1000 systems, and guess its more around 0.78, 1.2, and 1.95, but that is only for mature galaxies.

The answer is a lot more complicated. E.g. for high density and young, I get around 1.2 and for ancient 2.4. And then the number changes over time due to nebulae.

Reply:
```
          Low   Medium    High
Ancient  x0.92   x1.46   x2.42
Mature   x0.78   x1.17   x1.95
Young    x0.54   x0.79   x1.32
```
I run 5 repetitions for each combination of planet density and galaxy age. All tests where done with 1000 systems.
Then, for the Mature-Medium setting, I run 3 more repetitions for each galaxy shape (tested spiral 2-arms, cluster, disk, irregular and ring), the number of players (tested 7 and 21), and the starlane density, and I found no significant variations.

Reply:
Now I am looking at how to access those galaxy settings from the AI.

The code for universe generation `import freeorion as fo` and accesses `fo.get_galaxy_setup_data()` (and from there, it gets .size, .age, and .planetDensity). But AIs `import freeOrionAIInterface as fo`, and that interface has no `get_galaxy_setup_data` attribute. But the data I need is in that interface:
https://github.com/freeorion/freeorion/blob/691192a798ed1470b62e71be8fc1a6e6609d1de5/default/python/AI/freeOrionAIInterface.pyi#L46

Can anyone point me out in the right direction?


Also, I need a new table, something like this:
```
    age_to_density_factor = {
        #                                    Low  Medium  High
        #                        Young:
        fo.galaxySetupOptionGeneric.low:    (0.54, 0.79, 1.32),
        #                        Mature:
        fo.galaxySetupOptionGeneric.medium: (0.78, 1.17, 1.95),
        #                        Ancient:
        fo.galaxySetupOptionGeneric.high:   (0.92, 1.46, 2.42),
    }
```
It makes no sense to place it in the universe generation tables (it would not be used by universe generation at all), so for now I've placed it within the _rate_upkeep() function where it is used.

Reply:
Ok, I found it, it's just getGalaxySetupData().
The rest of my doubts I'll ask in the forum, since it's about python basics.